pkg/virtctl/create/vm.go

```diff
@@ -0,0 +1,581 @@
+ * limitations under the License.
+ *
+ * Copyright 2022 Red Hat, Inc.
+ *
+ */
+
+package create
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/spf13/cobra"
+	k8sv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+	v1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/api/instancetype"
+	cdiv1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
+	"sigs.k8s.io/yaml"
+
+	"kubevirt.io/kubevirt/pkg/virtctl/templates"
+)
+
+const (
+	VM = "vm"
+
+	NameFlag                   = "name"
+	RunStrategyFlag            = "run-strategy"
+	TerminationGracePeriodFlag = "termination-grace-period"
+	InstancetypeFlag           = "instancetype"
+	PreferenceFlag             = "preference"
+	DataSourceVolumeFlag       = "volume-datasource"
+	ContainerdiskVolumeFlag    = "volume-containerdisk"
+	ClonePvcVolumeFlag         = "volume-clone-pvc"
+	PvcVolumeFlag              = "volume-pvc"
+	BlankVolumeFlag            = "volume-blank"
+	CloudInitUserDataFlag      = "cloud-init-user-data"
+	CloudInitNetworkDataFlag   = "cloud-init-network-data"
+
+	cloudInitDisk = "cloudinitdisk"
+)
+
+type createVM struct {
+	name                   string
+	terminationGracePeriod int64
+	runStrategy            string
+	instancetype           string
+	preference             string
+	dataSourceVolumes      []string
+	containerdiskVolumes   []string
+	clonePvcVolumes        []string
+	blankVolumes           []string
+	pvcVolumes             []string
+	cloudInitUserData      string
+	cloudInitNetworkData   string
+}
+
+type cloneVolume struct {
+	Name   string             `param:"name"`
+	Source string             `param:"src"`
+	Size   *resource.Quantity `param:"size"`
+}
+
+type pvcVolume struct {
+	Name   string `param:"name"`
+	Source string `param:"src"`
+}
+
+type blankVolume struct {
+	Name string             `param:"name"`
+	Size *resource.Quantity `param:"size"`
+}
+
+type optionFn func(*createVM, *v1.VirtualMachine) error
+
+var optFns = map[string]optionFn{
+	RunStrategyFlag:          withRunStrategy,
+	InstancetypeFlag:         withInstancetype,
+	PreferenceFlag:           withPreference,
+	DataSourceVolumeFlag:     withDataSourceVolume,
+	ContainerdiskVolumeFlag:  withContainerdiskVolume,
+	ClonePvcVolumeFlag:       withClonePvcVolume,
+	PvcVolumeFlag:            withPvcVolume,
+	BlankVolumeFlag:          withBlankVolume,
+	CloudInitUserDataFlag:    withCloudInitData,
+	CloudInitNetworkDataFlag: withCloudInitData,
+}
+
+var flags = []string{
+	RunStrategyFlag,
+	InstancetypeFlag,
+	PreferenceFlag,
+	DataSourceVolumeFlag,
+	ContainerdiskVolumeFlag,
+	ClonePvcVolumeFlag,
+	PvcVolumeFlag,
+	BlankVolumeFlag,
+	CloudInitUserDataFlag,
+	CloudInitNetworkDataFlag,
+}
+
+var runStrategies = []string{
+	string(v1.RunStrategyAlways),
+	string(v1.RunStrategyManual),
+	string(v1.RunStrategyHalted),
+	string(v1.RunStrategyOnce),
+	string(v1.RunStrategyRerunOnFailure),
+}
```

Comment:
supernit - would be nice to have this provided by kubevirt itself in the future.